### PR TITLE
redirect to tab

### LINKED ISSUE #1911: edit: remember currently opened tab on save
after saving the page, we get a reload and end up on the first tab again. it would be nice to remember the currently opened tab and open on the same tab again after save.

/cc @benoitpointet

### PR DESCRIPTION
**Pull request by osavchenko**
I am targeting this branch, because it's a small feature.

Closes #1911

## Changelog
```markdown
### Added
- Added ability to back to the tab which was edited on saving an object
```

### CODE REVIEW

File: src/Resources/public/Admin.js [hunk 1]
```diff
@@ -758,6 +758,7 @@ var Admin = {
             window.history.pushState({
                 path: newurl
             }, '', newurl);
+            jQuery(this).closest('form').find('input[name="_tab"]').val(tab);
```
Comment (kunicmarko20) [inline src/Resources/public/Admin.js]: is `closest` really the best option?

Comment (osavchenko) [reply src/Resources/public/Admin.js]: As an alternative, we can add an id to the form and data attribute to the tabs.

File: src/Controller/CRUDController.php [hunk 2]
```diff
@@ -1272,6 +1272,20 @@ protected function redirectTo($object)
...
+            if ($tab) {
+                $query = parse_url($url, PHP_URL_QUERY);
+
+                if ($query) {
+                    $url .= '&';
+                } else {
+                    $url .= '?';
+                }
+
+                $url .= "_tab={$tab}";
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: Why are not using the third argument of `AbstractAdmin::generateObjectUrl` on line 1270 instead?

File: src/Controller/CRUDController.php [hunk 3]
```diff
@@ -1267,7 +1267,14 @@ protected function redirectTo($object)
...
                 if ($this->admin->hasRoute($route) && $this->admin->hasAccess($route, $object)) {
-                    $url = $this->admin->generateObjectUrl($route, $object);
+                    $tab = $request->request->get('_tab');
+                    $params = [];
+
+                    if ($tab) {
+                        $params['_tab'] = $tab;
+                    }
+
+                    $url = $this->admin->generateObjectUrl($route, $object, $params);
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: You can simplify even more:
```suggestion
                    $url = $this->admin->generateObjectUrl(
                        $route, 
                        $object, 
                        $request->request->has('_tab') ? ['_tab' => $request->request->get('_tab')] : []);
```

**Comment by greg0ire** [comment]
Please fix flint CI by remove the trailing whitespaces in the lines I contributed (sorry for that, and squash your commits together).

If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already pushed. Don't forget the `--force` option otherwise git will try to merge both things together.

File: src/Resources/public/Admin.js (hunk 1 above)
Comment (OskarStark) [reply src/Resources/public/Admin.js]: ping @kunicmarko20

Comment (kunicmarko20) [reply src/Resources/public/Admin.js]: that sounds like a better way, not sure how reliable is `closest` what if there are multiple forms

**Comment by OskarStark** [comment]
Can you please rebase your PR?

**Comment by osavchenko** [comment]
Sure, will do it today later

File: src/Resources/public/Admin.js (hunk 1 above)
Comment (greg0ire) [reply src/Resources/public/Admin.js]: @kunicmarko20 @osavchenko please resolve this conversation.

Comment (osavchenko) [reply src/Resources/public/Admin.js]: @kunicmarko20 I have changed behaviour in [new commit](https://github.com/sonata-project/SonataAdminBundle/pull/5314/commits/65e88ebf6df249cfa2b29f46ee343c4162b54bd7)

**Comment by core23** [comment]
Thanks @osavchenko

**Comment by franmomu** [comment]
Hi there, this is not working to me, there are some issues:

The most important is that if you open a form with tabs and you save it without changing tab, it redirects to: `x/edit?_tab=` which results in showing a blank tab. This is because it always adds the `_tab` parameter to the form, so the first time is empty: 

`<input type="hidden" name="_tab" value="{{ app.request.query.get('_tab') }}">`

And in the CRUDController: `$request->request->has('_tab')` returns `true`, so: `$request->request->has('_tab') ? ['_tab' => $request->request->get('_tab')] : []` returns `['_tab' => '']`. I guess this is easy to fix with: 

`'' !== $request->request->get('_tab') ? ['_tab' => $request->request->get('_tab')] : []` returns `['_tab' => '']`

This code `jQuery(this).parent('.nav-tabs-custom').find('input[name="_tab"]').val(tab);` doesn't work because `.nav-tabs-custom` is not the direct parent of the tabs, I guess `closest` could be used.

One thing I particularly don't like is that changing the form dynamically clicking on the tabs (modifying the _tab value), means that if you then click to another item of the menu for example, it will warn you that the form has changed and the changes may not be saved (at least in Chrome).

**Pull request merged**